src/parser/parser_env.ml

```diff
@@ -357,12 +344,20 @@ let add_used_private env name loc =
 let consume_comments_until env pos = env.consumed_comments_pos := pos
 
 (* lookahead: *)
-let lookahead ~i env =
-  assert (i < maximum_lookahead);
-  Lookahead.peek !(env.lookahead) i
+let[@inline] lookahead_0 env = Lookahead.peek_0 !(env.lookahead)
+
+let[@inline] lookahead_1 env = Lookahead.peek_1 !(env.lookahead)
+
+let[@inline] lookahead ~i env =
+  match i with
+  | 0 -> lookahead_0 env
+  | 1 -> lookahead_1 env
+  | _ -> assert false
 
 (* functional operations: *)
-let with_strict in_strict_mode env = { env with in_strict_mode }
+let with_strict in_strict_mode env = 
+  if in_strict_mode = env.in_strict_mode then env 
```

Comment:
If we're going to do this optimization, it might be nice to do it uniformly for all `with_foo` combinators. I imagine you made the changes based on what bubbled up from profiling, yeah? That seems like it risks overfitting your benchmark. If this is a useful optimization, let's apply it to the rest of these functions as well.

Reply:
>  I imagine you made the changes based on what bubbled up from profiling, yeah?

Yes.
> If this is a useful optimization, let's apply it to the rest of these functions as well.

Done in coming commits